arrowload stata command to run python script

rebkwok opened the pull request.

rebkwok commented:
I've tested this reading .feather output from cohort-extractor and .arrow output from databuilder

rebkwok commented:
There's a bug that I've just noticed (I've been trying this with arrow files from the initiative 1 studies).
Stata has a limit of 32 characters for variable names, and the `sfi.SFIToolkit.strToName` method I've used to clean the names just truncates them if they're too long.  So when a study has lots of iterations over visits/events etc, and has variable names like `is_opted_out_of_nhs_data_share_1`, `is_opted_out_of_nhs_data_share_2` etc the name-cleaning results in duplicate column names, which means it drops some columns.

I'm not sure what the best solution is - it could raise an error if the name-cleaning results in duplicate names, or if any names are >32 characters, and just prompt users to rename their variables to shorter names.  Or I can do something to generate unique names, by truncating them in some way and adding an incremental suffix, but depending on how the variables were named in the first place, that could result in variables that aren't easily matched up to the ones in the input file.

rebkwok commented:
This is now ready for review again:
- made the command take optional arguments properly
- option to load an aliases csv file and errors for invalid variable names
- added a help file

bloodearnest commented on python_scripts/load_arrow.py:
```diff
@@ -1,13 +1,295 @@
+                sfi.Data.addVarInt(varname)
+            elif vartype == "category":
+                sfi.Data.addVarLong(varname)
+            elif vartype in ["float", "date", "timestamp"]:
```
dates and timestamps are floats?!

oh my...

rebkwok replied:
Dates and timestamps are numeric in stata (afaict we only have to deal with dates, so days since 1st Jan 1960 is what we need to convert to, but they could be seconds), and b/c of the int size limit, they need to be float type. I think. I'm still somewhat vague on this.

bloodearnest commented on python_scripts/load_arrow.py:
```diff
@@ -0,0 +1,348 @@
+                print(f"{varname} aliased to {cleaned_name}")
+            clean_column_names.append(cleaned_name)
+
+        if too_long_names:
```
If the too_long_names can be automatically shortened w/o clashing, should we do that? And print a warning that we have done it?

I assume stata users are used to working withing the 32 character limit anyway? Is there some form of autoshortening that stata provides when loading from csv we can copy? e.g. "`too_long_name_xxxxx" -> "to_long_na~1"` or similar?

rebkwok replied:
They could be automatically shortened, but we decided not to, and raising exceptions with warnings was the better route - see [this slack thread](https://bennettoxford.slack.com/archives/C03FB777L1M/p1681229679397389) that I totally forgot was needed for context

bloodearnest commented on libraries/arrowload.sthlp:
Custom stata specific help! :clap: :clap: :clap: 

I salute you!

bloodearnest commented on tests/run.sh:
```diff
@@ -79,8 +78,28 @@ assert-content "data was ungzipped" "tests/output/input.csv" "1,2,3"
+# - int: `age`, `patient_id` (int64 in feather file, int in stata)
+# - categorical: sex (coded M/F as 0/1, no nulls), region (includes nulls)
+# - num_val: int64 in feather file (values with mean 50,000), converted to long in stata
+try "load arrow file" analysis/arrowload.do
```
Ugh, sorry about the custom bash test suite. We previously didn't have a python venv, but now we do, we should probably rewrite this in pytest.

rebkwok replied:
I thought about it, but wasn't sure how we'd write pytest tests for something that required the stata sfi module.  Not that I tried it though

bloodearnest reviewed (approved):
This. Is. Awesome.

Excellent job diving into an unfamiliar ecosystem and grasping the nettle.  I continue to be impressed by your ability to grapple with new problems and ship solutions. Bravo!


Left a couple of nit comments, and one possible suggestion for the 32 column limit issue, but happy to land as is.

Appreciate the project infra changes too.

rebkwok commented on python_scripts/load_arrow.py:
```diff
@@ -0,0 +1,348 @@
+        Function used to create the Stata variables in memory.
+        This assumes that the values in name_and_type object are valid
+        Stata variable names.
+        :param name_and_type: A dictionary structured to contain the variable names
```
These are leftovers, I'll get rid of/update them

rebkwok commented on Dockerfile:
```diff
@@ -9,7 +9,71 @@
+
+# we are going to use an apt cache on the host, so disable the default debian
+# docker clean up that deletes that cache on every apt install
+RUN rm -f /etc/apt/apt.conf.d/docker-clean
```
:+1: I was copy/pasting from some other Dockerfile, can't you tell?

evansd reviewed (commented):
As Simon says, this really is awesome. I don't think we're that far off being done but I have made quite a few detailed comments, mainly because the closer you look at Stata the more weirdness you discover. I probably haven't been that clear in some of them so very happy to discuss further if needed.

rebkwok commented on Dockerfile:
```diff
@@ -4,12 +4,71 @@
+
+# copy /root/.local files over from builder image to get the installed python dependencies.
+# These will have root:root ownership, but are readable by all.
+COPY --from=builder /root/.local/lib/python3.11/site-packages /usr/local/lib/python3.11/site-packages
```
I'll give that a try.  Getting them to be recognised at all was a bit of a trial-and-error struggle.

rebkwok commented on python_scripts/load_arrow.py:
```diff
@@ -1,13 +1,295 @@
+                sfi.Data.addVarInt(varname)
+            elif vartype == "category":
+                sfi.Data.addVarLong(varname)
+            elif vartype in ["float", "date", "timestamp"]:
```
It does, but it also can't deal with big ints, so if there's a timestamp that becomes a very big number when we convert it to an int, I think it would need to be a float.  Timestamps can be stored as number of milliseconds since 1st Jan 1960, and I don't think stata can cope with that as a integer type.

evansd replied:
So that's definitely true for dates and at the moment dates are all we support in ehrQL so we should be able to store those as `int32` and everything should be fine.

But looking again at timestamps, I think Stata's docs are just internally inconsistent here:
![image](https://user-images.githubusercontent.com/19630/232725971-c2641b26-4fb0-4543-ac79-782f8bb54138.png)

The example integer they give is bigger than the biggest integer type you can store in Stata. So I think timestamps must be stored as floats, with the assumption that the loss of precision doesn't matter for the purposes for which they're used.

So I guess we have to store timestamps as floats. Or we could punt on the issue, given that we don't yet emit timestamps from ehrQL.

rebkwok commented on python_scripts/load_arrow.py:
```diff
@@ -1,13 +1,295 @@
+class ArrowConverter:
+    def __init__(self, filename, max_chunksize):
+        # read the feather file into a table
+        print(f"Reading file {filename}")
```
Print statements go into the stata output and log

rebkwok commented on python_scripts/load_arrow.py:
```diff
@@ -1,13 +1,295 @@
+                if self.column_types[varname] == "category":
+                    # convert category data to a list of indices
+                    column_data = (
+                        batch.columns[i].dictionary_encode().indices.to_pylist()
```
I'm pretty sure I tried that and you can't, but I'll check

rebkwok commented on python_scripts/load_arrow.py:
```diff
@@ -1,13 +1,295 @@
+                sfi.Data.addVarStrL(varname)
+            elif vartype == "boolean":
+                sfi.Data.addVarByte(varname)
+            elif vartype == "int":
+                sfi.Data.addVarInt(varname)
```
@evansd I am looking at this now, and it is going OK, except for categoricals with missing values.  It's pretty straightforward to find the min/max and then the appropriate Stata type for those.  And for categoricals, we find the size of the dictionary and use that and 0 for min/max, since the values are indices.

But, when a categorical has null values, and we replace that null value in the indices list with the appropriate stata null value (101 for bytes, which is what most categoricals will be), it gets converted to `int` when it's loaded into stata (even though the variable was created as a byte-type var).  Similarly, if I make all categoricals ints, and replace nulls with the int missing value (32768), it gets converted to `long` when it's loaded into stata. Only when we have float/double and the missing value is the stata sysmiss do they seem to get properly encoded as missing when they get into stata.

Any ideas?

evansd replied:
This is all a bit depressing, isn't it?

The only thing I can think of it is to try to create a categorical variable with missing values from within Stata and then try to reverse engineer what it's done. Possibly by writing it out as a `.dta` file and then loading it into Pandas? Or – worst case – by hex-spelunking the `.dta` file itself to work out how it's been encoded.

Or maybe it will turn out that Stata doesn't really handle missing values in categoricals. We might need to invoke Alex as a our resident Stata person here.

rebkwok replied:
Actually I think I was wrong - it's not just categoricals.  If there's an int, and I create a Int-type stata variable for it, it automatically loads it to stata as `long` and treats the null-replacement value as ints.  One thought I had was to load it, and then replace the null-replacement values in stata based on the types we originally define (pre-null replacement)

evansd replied:
Ah, right. That almost makes more sense, not that it helps us much!

I'm aware that I've been chucking out grand plans in the abstract based on how I _think_ it ought to work without knowing how it actually does work.

Does that mean there's just no way of passing missing values up through the Python interface? That's a bit unexpected, but I suppose it wouldn't be entirely surprising. What happens with Python `None` values?

rebkwok replied:
Python None values (if you don't replace them with something) make it error when you try to load it to stata

evansd replied:
I don't fully understand what this documentation is saying, but is it possible that we could do something with the `selectvar` argument here? Will setting it to `-1` somehow possibly do what we want?
https://www.stata.com/python/api16/Data.html#sfi.Data.store

rebkwok replied:
I already tried that, and I don't think so. I couldn't get it to work, I got this error, no matter what I tried.
```
Traceback (most recent call last):
  File "/python_scripts/load_arrow.py", line 455, in <module>
    main(**parse_args())
  File "/python_scripts/load_arrow.py", line 23, in main
    converter.load_data()
  File "/python_scripts/load_arrow.py", line 432, in load_data
    sfi.Data.store(
  File "/usr/local/stata/ado/base/py/sfi.py", line 1431, in store
    return _stp._st_storedata(vars, obss, val, svar)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SystemError: conformability error; failed to store values in the current Stata 
```
(That's the same error I get if I leave the None values)

I also tried with just one column, in case the -1 option required named variables in `var`, but same error.

I have a solution now, which is to load it into stata and then run stata commands to replace the numeric null value with `.` and cast the variable to what we determined it should be in the first place.

rebkwok commented:
@evansd I think this is ready for review again. I've addressed all of your comments, and done some more re-working after I made some more comprehensive test files and found bugs and slow things (esp the date/timestamp conversions).  So I'm afraid it pretty much needs an entire re-review.  I've kept all the incremental commits, which I think make reasonably logical sense.

evansd commented on python_scripts/load_arrow.py:
```diff
@@ -379,7 +379,8 @@ def make_vars(self, pre_processed_column_types):
             for varname, vartype in self.column_types.items():
                 if vartype == "string":
-                    sfi.Data.addVarStrL(varname)
+                    max_length = max(len(val.as_py()) for (val) in batch[varname])
```
I guess this has the same potential edge case as with numeric ranges where the first batch might not be representative. But again, I think it's unlikely and as long as the failure more is an explicit error rather than silent corruption then I think it's OK.

rebkwok replied:
I'm not sure this is explicitly checked in tests, but I _think_ stata just recasts strings if they're not big enough - I'll look at adding a test in my pytest PR to follow :)

evansd reviewed (approved):
Well haven't you been having fun!

I think this is good to go. I had some questions about what happens when ranges/string lengths in the first batch aren't representative, but I think maybe the final commit answers this because we recast if necessary after each batch.
